### PR TITLE
Feature/tap regulator output

### PR DESCRIPTION
**Pull request by mgovers**
output tap positions for the regulators with the original implementation of #598

### CODE REVIEW

**Comment by mgovers** [comment]
@nitish let's discuss tomorrow whether to merge this into #598 and/or also to `main`. that would mean we reach the sprint goal and cleanup (making it efficient) would then be a follow-up

File: power_grid_model_c/power_grid_model/include/power_grid_model/calculation_parameters.hpp [hunk 1]
```diff
@@ -127,7 +127,7 @@ struct TransformerTapRegulatorCalcParam {
 };
 
 struct TransformerTapPosition {
-    Idx2D index{};
+    ID transformer{};
```
Comment (Jerry-Jinfeng-Guo) [inline power_grid_model_c/power_grid_model/include/power_grid_model/calculation_parameters.hpp:130]:
```suggestion
    ID transformer_id{};
```

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 2]
```diff
@@ -622,21 +610,24 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
...
         TransformerTapPositionResult transformer_tap_positions;
 
-        // TODO(mgovers): only output the transformers that are regulated
-        (get_transformer_tap_positions<TransformerTypes, typename State::ComponentContainer>(state,
-                                                                                             transformer_tap_positions),
-         ...);
+        for (auto const& sub_order : regulator_order) {
+            for (auto const& regulator : sub_order) {
+                auto const& transformer = regulator.transformer;
+                transformer_tap_positions.push_back({transformer.id(), transformer.tap_pos()});
```
Comment (Jerry-Jinfeng-Guo) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:620]:
```suggestion
                transformer_tap_positions.emplace_back({transformer.id(), transformer.tap_pos()});
```
on behalf of sonar cloud

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 3]
```diff
@@ -622,21 +610,24 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
...
+    auto produce_output(std::vector<std::vector<RegulatedTransformer>> const& regulator_order,
+                        ResultType solver_output) const -> MathOutput<ResultType> {
         TransformerTapPositionResult transformer_tap_positions;
 
-        // TODO(mgovers): only output the transformers that are regulated
-        (get_transformer_tap_positions<TransformerTypes, typename State::ComponentContainer>(state,
-                                                                                             transformer_tap_positions),
-         ...);
+        for (auto const& sub_order : regulator_order) {
+            for (auto const& regulator : sub_order) {
```
Comment (Jerry-Jinfeng-Guo) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:618]: Sooo a nested `for` then

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 4]
```diff
@@ -990,47 +1006,6 @@ TEST_CASE("Test tap position optmizer I/O") {
                                                                           bad_regulators.end(), 50.0),
                         DuplicativelyRegulatedObject);
     }
-
-    SUBCASE("transformer tap position retrieval") {
```
Comment (Jerry-Jinfeng-Guo) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:994]: So the decision is to not unit test tap pos retrieval?

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 2 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:620]: i now know why you initially used `push_back`. see https://developercommunity.visualstudio.com/t/stdvectoremplace-back-does-not-work-for-a-class-wi/1691516 and in particular this minimal example: https://godbolt.org/z/xovjh9ffe . 

looks like a compiler bug. we can change it in the future when it's resolved

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:620]: FYI: this was fixed in clang `16.0.0` https://godbolt.org/z/vYohvbv7q

Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:620]: another reason to migrate to Ubuntu 24.04 runner images

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp (hunk 3 above)
Comment (mgovers) [reply power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:618]: it is what it is. it's fine, i guess. it is definitely more readable than more variadic stuff (that is handled internally in the transformer regulator reference class)

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp (hunk 4 above)
Comment (mgovers) [reply tests/cpp_unit_tests/test_tap_position_optimizer.cpp:994]: it should be unit tested in `main_core` but here, the class' unit tests itself should take care of it. see https://github.com/PowerGridModel/power-grid-model/pull/607/files#diff-e3cb2a7b8db22693560826087f2c675630f9cd51a2337191aadc986dd50c4249R951-R958

**Pull request merged**